server/src/main/java/org/opensearch/rest/RequestLimitSettings.java

```diff
@@ -0,0 +1,144 @@
+    /**
+     * Enum to represent action names against whom we need to perform limit checks.
+     */
+    public enum BlockAction {
+        CAT_INDICES,
+        CAT_SHARDS,
+        CAT_SEGMENTS
+    }
+
+    private volatile int catIndicesLimit;
+    private volatile int catShardsLimit;
+    private volatile int catSegmentsLimit;
+
+    /**
+     * Setting to enable circuit breaker on {@link RestIndicesAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_INDICES_LIMIT_SETTING = Setting.intSetting(
+        "cat.indices.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable circuit breaker on {@link RestShardsAction}. The limit will be applied on number of shards.
+     */
+    public static final Setting<Integer> CAT_SHARDS_LIMIT_SETTING = Setting.intSetting(
+        "cat.shards.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable circuit breaker on {@link RestSegmentsAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_SEGMENTS_LIMIT_SETTING = Setting.intSetting(
+        "cat.segments.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public RequestLimitSettings(ClusterSettings clusterSettings, Settings settings) {
+        setCatShardsLimitSetting(CAT_SHARDS_LIMIT_SETTING.get(settings));
+        setCatIndicesLimitSetting(CAT_INDICES_LIMIT_SETTING.get(settings));
+        setCatSegmentsLimitSetting(CAT_SEGMENTS_LIMIT_SETTING.get(settings));
+
+        clusterSettings.addSettingsUpdateConsumer(CAT_SHARDS_LIMIT_SETTING, this::setCatShardsLimitSetting);
+        clusterSettings.addSettingsUpdateConsumer(CAT_INDICES_LIMIT_SETTING, this::setCatIndicesLimitSetting);
+        clusterSettings.addSettingsUpdateConsumer(CAT_SEGMENTS_LIMIT_SETTING, this::setCatSegmentsLimitSetting);
+    }
+
+    /**
+     * Method to check if the circuit breaker limit has reached for an action.
+     * The limits are controlled via dynamic settings.
+     *
+     * @param clusterState  {@link ClusterState}
+     * @param actionToCheck {@link BlockAction}
+     * @return True/False
+     */
+    public boolean isCircuitLimitBreached(final ClusterState clusterState, final BlockAction actionToCheck) {
+        if (Objects.isNull(clusterState)) return false;
+        switch (actionToCheck) {
+            case CAT_INDICES:
+                if (catIndicesLimit <= 0) return false;
+                int indicesCount = chainWalk(() -> clusterState.getMetadata().getIndices().size(), 0);
+                if (indicesCount > catIndicesLimit) return true;
+                break;
+            case CAT_SHARDS:
+                if (catShardsLimit <= 0) return false;
+                final RoutingTable routingTable = clusterState.getRoutingTable();
```

Comment:
Any specific reason to get number of shards from routing table only and not from ```ClusterState.metadata```? I think we might also have to provide a way to override this check, i.e. overriding RestAction's ```Predicate<ClusterState>``` if provided.

The thing is either of routingTable or metadata object can be empty on the basis of what was queried for clusterState. I think as of today, CatShards is not requesting for metadata object, but some other action might not be requesting for RoutingTable and then this default check could fail again.

Reply:
We were not having the metadata populated in case of Cat Shards that why we had to use routing table. 

With the new changes of introducing BlockEntity, i have refactored the code and kept things simple. We will call isCircuitLimitBreached method with metadata or routing table depending on the use-case.